bolt-contracts/src/contracts/BoltValidatorsV2.sol

```diff
@@ -0,0 +1,295 @@
+        }
+
+        _batchRegisterValidators(pubkeyHashes, authorizedOperator, maxCommittedGasLimit);
+    }
+
+    // ========= UPDATE FUNCTIONS =========
+
+    /// @notice Update the maximum gas limit that a validator can commit for preconfirmations
+    /// @dev Only the `controller` of the validator can update this value.
+    /// @param pubkeyHash The hash of the BLS public key of the validator
+    /// @param maxCommittedGasLimit The new maximum gas limit
+    function updateMaxCommittedGasLimit(bytes20 pubkeyHash, uint32 maxCommittedGasLimit) public {
+        address controller = VALIDATORS.getController(pubkeyHash);
+        if (msg.sender != controller) {
+            revert UnauthorizedCaller();
+        }
+
+        VALIDATORS.updateMaxCommittedGasLimit(pubkeyHash, maxCommittedGasLimit);
+    }
+
+    // ========= HELPERS =========
+
+    /// @notice Internal helper to register a single validator
+    /// @param pubkeyHash BLS public key hash of the validator
+    /// @param authorizedOperator Address of the authorized operator
+    /// @param maxCommittedGasLimit Maximum gas limit that the validator can commit for preconfirmations
+    function _registerValidator(bytes20 pubkeyHash, address authorizedOperator, uint32 maxCommittedGasLimit) internal {
+        if (authorizedOperator == address(0)) {
+            revert InvalidAuthorizedOperator();
+        }
+        if (pubkeyHash == bytes20(0)) {
+            revert InvalidPubkey();
+        }
+
+        VALIDATORS.insert(
+            pubkeyHash,
+            maxCommittedGasLimit,
+            VALIDATORS.getOrInsertController(msg.sender),
+            VALIDATORS.getOrInsertAuthorizedOperator(authorizedOperator)
+        );
+        emit ValidatorRegistered(pubkeyHash);
+    }
+
+    /// @notice Internal helper to register a batch of validators
+    /// @param pubkeyHashes List of BLS public key hashes of the validators
+    /// @param authorizedOperator Address of the authorized operator
+    /// @param maxCommittedGasLimit Maximum gas limit that the validators can commit for preconfirmations
+    function _batchRegisterValidators(
+        bytes20[] memory pubkeyHashes,
+        address authorizedOperator,
+        uint32 maxCommittedGasLimit
+    ) internal {
+        if (authorizedOperator == address(0)) {
+            revert InvalidAuthorizedOperator();
+        }
+
+        uint32 authorizedOperatorIndex = VALIDATORS.getOrInsertAuthorizedOperator(authorizedOperator);
+        uint32 controllerIndex = VALIDATORS.getOrInsertController(msg.sender);
+        uint256 pubkeysLength = pubkeyHashes.length;
+
+        for (uint32 i; i < pubkeysLength; i++) {
+            bytes20 pubkeyHash = pubkeyHashes[i];
+
+            if (pubkeyHash == bytes20(0)) {
+                revert InvalidPubkey();
+            }
+
+            VALIDATORS.insert(pubkeyHash, maxCommittedGasLimit, controllerIndex, authorizedOperatorIndex);
+            emit ValidatorRegistered(pubkeyHash);
```

Comment:
Is it really necessary to emit an event in every iteration? Since this unsafe registration the public keys might well be invalid